Fix Typos and Styling for LimitRange Documentation (#18961)
kubernetes/website: content/en/docs/concepts/policy/limit-range.md

---
reviewers:

> - nelvadas

Reviewer: The original title, “Limit Ranges”, should localize into target languages. I like that it would.
How about changing the title to “Resource Limit Ranges”?

Author: I guess I would prefer to just keep `Limit Ranges` in this case, but I agree on the rationale.

title: LimitRanges
content_template: templates/concept
weight: 10
---

{{% capture overview %}}

> By default, Containers run with unbounded [compute resources](/docs/user-guide/compute-resources) on a Kubernetes cluster.

Reviewer: I prefer the original line. The term “container” here is about the general IT concept rather than an element of Kubernetes' API.

Author: Sure, I will return to lowercase. My fault. I didn't realize capitals were reserved for API concepts.

With resource quotas, cluster administrators can restrict resource consumption and creation on a namespace basis.

> Within a namespace, a Pod or Container can consume as much CPU and memory as defined by the namespace's resource quota. There is a concern that one Pod or Container could monopolize all available resources. A LimitRange is a policy to constrain resource consumption by Pod or Container in a namespace.

Reviewer: How about: “A LimitRange is a policy to constrain resource allocations (to Pods or Containers) in a namespace.”

One thing I want readers to learn is that LimitRanger can enforce _minimum_ request (and _minimum_ limit). So if I want to create a Pod that requests 5m CPU and 16KiB memory, someone who manages that namespace can force me to request more CPU and more memory. I'm thus restricted in my ability to mislead the scheduler.

{{% /capture %}}

{{% capture body %}}

> A LimitRange, defined by a `LimitRange` object, provides constraints that can:

Reviewer: This seems tautological. How about:
```suggestion
A _LimitRange_ provides constraints that can:
```

- Enforce minimum and maximum compute resources usage per Pod or Container in a namespace.
- Enforce minimum and maximum storage request per PersistentVolumeClaim in a namespace.
- Enforce a ratio between request and limit for a resource in a namespace.
- Set default request/limit for compute resources in a namespace and automatically inject them to Containers at runtime.

## Enabling LimitRange

LimitRange support is enabled by default for many Kubernetes distributions. It is
enabled when the apiserver `--enable-admission-plugins=` flag has `LimitRanger` admission controller as
one of its arguments.

A LimitRange is enforced in a particular namespace when there is a

> `LimitRange` object in that namespace.

Reviewer: Let's also fix this:
```suggestion
LimitRange object in that namespace.
```

### Overview of LimitRange:

- The administrator creates one `LimitRange` in one namespace.
- Users create resources like Pods, Containers, and PersistentVolumeClaims in the namespace.

> - The `LimitRanger` admission controller enforces defaults and limits for all Pods and Containers that do not set compute resource requirements and tracks usage to ensure it does not exceed resource minimum, maximum and ratio defined in any `LimitRange` present in the namespace.

Reviewer: ```suggestion
- The `LimitRanger` admission controller enforces defaults and limits for all Pods and Containers that do not set compute resource requirements and tracks usage to ensure it does not exceed resource minimum, maximum and ratio defined in any LimitRange present in the namespace.
```
LimitRange is a resource.

  requests or limits for those values. Otherwise, the system may reject pod creation.
- LimitRange validations occurs only at Pod Admission stage, not on Running Pods.

Examples of policies that could be created using a LimitRange are:

- In a 2 node cluster with a capacity of 8 GiB RAM and 16 cores, constrain Pods in a namespace to request 100m of CPU with a max limit of 500m for CPU and request 200Mi for Memory with a max limit of 600Mi for Memory. 
- Define default CPU limit and request to 150m and memory default request to 300Mi for containers started with no cpu and memory requests in their specs.

In the case where the total limits of the namespace is less than the sum of the limits of the Pods/Containers,
there may be contention for resources. In this case, the Containers or Pods will not be created.

Neither contention nor changes to a LimitRange will affect already created resources.

## Limiting Container compute resources

The following section discusses the creation of a LimitRange acting at Container Level.
A Pod with 04 containers is first created. Each container within the Pod has a specific `spec.resource` configuration.

> Each container within the pod is handled differently by the LimitRanger admission controller.

Reviewer: ```suggestion
Each container within the pod is handled differently by the `LimitRanger` admission controller.
```

{{% capture overview %}}

By default, containers run with unbounded [compute resources](/docs/user-guide/compute-resources) on a Kubernetes cluster.
With resource quotas, cluster administrators can restrict resource consumption and creation on a namespace basis.

> Within a namespace, a Pod or Container can consume as much CPU and memory as defined by the namespace's resource quota. There is a concern that one pod or container could monopolize all available resources. A Limit Range is a policy to constrain resource consumption by Pod or Container in a namespace.

Reviewer: Did you mean:
```suggestion
Within a namespace, a Pod or Container can consume as much CPU and memory as defined by the namespace's resource quota. There is a concern that one pod or container could monopolize all available resources. A LimitRange is a policy to constrain resource consumption by Pod or Container in a namespace.
```

or maybe:
```suggestion
Within a namespace, a Pod or Container can consume as much CPU and memory as defined by the namespace's resource quota. There is a concern that one pod or container could monopolize all available resources. A _limit range_ is a policy to constrain resource consumption by Pod or Container in a namespace.
```
?

Author: Yeah, I think that's the question I have is. It's referenced several ways in the doc. I think `LimitRange`, `Limit Range`, or `limit range` would be acceptable, but I'll review the style guide you have provided.

- The administrator creates one `LimitRange` in one namespace.
- Users create resources like Pods, Containers, and PersistentVolumeClaims in the namespace.

> - The `LimitRanger` admission controller enforces defaults and limits for all Pods and Container that do not set compute resource requirements and tracks usage to ensure it does not exceed resource minimum, maximum and ratio defined in any `LimitRange` present in the namespace.

Reviewer: To follow the style guidelines, this could become:
```suggestion
- The `LimitRanger` admission controller enforces defaults and limits for all Pods and Containers that do not set compute resource requirements and tracks usage to ensure it does not exceed resource minimum, maximum and ratio defined in any LimitRange present in the namespace.
```

- If creating or updating a resource (Pod, Container, PersistentVolumeClaim) that violates a limit range  constraint, the request to the API server will fail with an HTTP status code `403 FORBIDDEN` and a message explaining the constraint that have been violated.
- If a limit range is activated in a namespace for compute resources like `cpu` and `memory`, users must specify
  requests or limits for those values. Otherwise, the system may reject pod creation.
- Limit Range validations occurs only at Pod Admission stage, not on Running pods.

Examples of policies that could be created using limit range are:

> 

Author: The portion of the example here for `request 200Mi and not exceed 600Mi` did not make sense so I removed it. Is it referring to memory specifications for the limit range?

Reviewer: > request 200Mi and not exceed 600Mi

I take that to mean that Pods will have a 200MiB memory request and to require that matched Pods have a memory limit at or below 600MiB.

Author: Thanks! Let me add it back/make it more clear.

- In a 2 node cluster with a capacity of 8 GiB RAM and 16 cores, constrain Pods in a namespace to request 100m of CPU with a max limit of 500m for CPU
- Define default CPU limit and request to 150m and memory default request to 300Mi for containers started with no cpu and memory requests in their specs.

In the case where the total limits of the namespace is less than the sum of the limits of the Pods/Containers,
there may be contention for resources. In this case, the Containers or Pods will not be created.

Neither contention nor changes to a limit range will affect already created resources.

## Limiting Container compute resources

> The following section discusses the creation of a Limit Range acting at Container Level.

Reviewer: The resource is called LimitRange; the original line is correct.

```

> - The `busybox-cnt01` Container inside `busybox` Pod defined `requests.cpu=100m` and `requests.memory=100Mi`.

Reviewer: I'm happy with the existing “limit range” wording here.

Author: So question on this: I don't see limit range as two separate words in other docs on the website. It always seems to be LimitRange. Why would this be an exception?

Reviewer: Here, “limit range” is used to mean “a range of valid values for some subfield(s) of `requests` or `limits`” whereas elsewhere “LimitRange” means “the Kubernetes API object that represents a constraint on acceptable values for some subfield(s) of `requests` or `limits`” .

The name LimitRange is kind of unhelpful because it implies it's a constraint specifically on `limits`, rather than being applicable to both `requests` and `limits`.

Author: Ah, tricky! I agree, and it makes sense. I'll change back this and the others you have highlighted. Thank you.

-  `100m <= 500m <= 800m` , The Container cpu limit (500m) falls inside the authorized CPU LimitRange.
-  `99Mi <= 200Mi <= 1Gi` , The Container memory limit (200Mi) falls inside the authorized Memory LimitRange.

> -  No request/limits ratio validation for CPU/Memory, so the Container is valid and will be created.

Reviewer: Can you rewrite this without “will be”? Usually that's possible, and more in line with the style guide.

(You might have to change more of the text to achieve that).

Author: Removed it.

}
```

> -  The `busybox-cnt02` Container inside `busybox1` Pod defined `requests.cpu=100m` and `requests.memory=100Mi` but not limits for cpu and memory.

Reviewer: I'm happy with the existing “limit range” wording here.

```

> -  The `busybox-cnt03` Container inside `busybox1` Pod defined `limits.cpu=500m` and `limits.memory=200Mi` but no `requests` for cpu and memory.

Reviewer: I'm happy with the existing “limit range” wording here.

## Limits/Requests Ratio

If `LimitRangeItem.maxLimitRequestRatio` is specified in the `LimitRangeSpec`, the named resource must have a request and limit that are both non-zero where limit divided by request is less than or equal to the enumerated value.

> The following LimitRange enforces memory limit to be at most twice the amount of the memory request for any pod in the namespace:

Reviewer: nit, how about:
```suggestion
The following LimitRange enforces memory limit to be at most twice the amount of the memory request for any Pod in the namespace:
```
?

Author: Sorry, missed this one.

Pod         memory    -    -    -                -              2
```

Create a pod with `requests.memory=100Mi` and `limits.memory=300Mi`:

{{< codenew file="admin/resource/limit-range-pod-3.yaml" >}}

```shell
kubectl apply -f https://k8s.io/examples/admin/resource/limit-range-pod-3.yaml
```

The pod creation failed as the ratio here (`3`) is greater than the enforced limit (`2`) in `limit-memory-ratio-pod` LimitRange:

> ```shell

Reviewer: Aside: this is output, it shouldn't be marked as shell.

@danielhelfand it's up to you if you want to fix that or leave it to a different PR.

Author: I have updated it.
